iseq.c

```diff
@@ -248,230 +245,155 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
 
-        rb_iseq_each_value(iseq, each_insn_value, NULL);
+        rb_iseq_mark_and_move_each_value(iseq, reference_updating ? ISEQ_ORIGINAL_ISEQ(iseq) : NULL);
 
-        rb_gc_mark_movable(body->variable.coverage);
-        rb_gc_mark_movable(body->variable.pc2branchindex);
-        rb_gc_mark_movable(body->variable.script_lines);
-        rb_gc_mark_movable(body->location.label);
-        rb_gc_mark_movable(body->location.base_label);
-        rb_gc_mark_movable(body->location.pathobj);
-        RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)body->mandatory_only_iseq);
-        RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)body->parent_iseq);
+        rb_gc_mark_and_move(&body->variable.coverage);
+        rb_gc_mark_and_move(&body->variable.pc2branchindex);
+        rb_gc_mark_and_move(&body->variable.script_lines);
+        rb_gc_mark_and_move(&body->location.label);
+        rb_gc_mark_and_move(&body->location.base_label);
+        rb_gc_mark_and_move(&body->location.pathobj);
+        if (body->local_iseq) rb_gc_mark_and_move_ptr(&body->local_iseq);
+        if (body->parent_iseq) rb_gc_mark_and_move_ptr(&body->parent_iseq);
+        if (body->mandatory_only_iseq) rb_gc_mark_and_move_ptr(&body->mandatory_only_iseq);
 
         if (body->call_data) {
-            struct rb_call_data *cds = (struct rb_call_data *)body->call_data;
-            for (unsigned int i=0; i<body->ci_size; i++) {
-                const struct rb_callinfo *ci = cds[i].ci;
-                const struct rb_callcache *cc = cds[i].cc;
+            for (unsigned int i = 0; i < body->ci_size; i++) {
+                struct rb_call_data *cds = body->call_data;
 
-                if (vm_ci_markable(ci)) {
-                    rb_gc_mark_movable((VALUE)ci);
+                if (reference_updating) {
+                    if (!SPECIAL_CONST_P((VALUE)cds[i].ci)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].ci);
+                    }
+                }
+                else {
+                    if (vm_ci_markable(cds[i].ci)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].ci);
+                    }
                 }
 
-                if (cc) {
-                    VM_ASSERT((cc->flags & VM_CALLCACHE_ON_STACK) == 0);
+                if (cds[i].cc && (reference_updating || vm_cc_markable(cds[i].cc))) {
+                    VM_ASSERT(reference_updating || (cds[i].cc->flags & VM_CALLCACHE_ON_STACK) == 0);
 
-                    if (vm_cc_markable(cc)) {
-                        if (!vm_cc_invalidated_p(cc)) {
-                            rb_gc_mark_movable((VALUE)cc);
-                        }
-                        else {
-                            cds[i].cc = rb_vm_empty_cc();
-                        }
+                    if (reference_updating || !vm_cc_invalidated_p(cds[i].cc)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].cc);
+                    }
+                    else {
+                        cds[i].cc = rb_vm_empty_cc();
```

Comment:
This part of the code seems to have the semantics of "if `cds[i].cc` points to an invalidated object, then clear the `&cds[i].cc`field during GC".  Strictly speaking, a GC thread is not a mutator, and by definition a GC thread should not modify the object graph.  Re-assigning a field to `rb_vm_empty_cc()` is a form of modifying the object graph.

I think it has similar semantics as weak references.  And the "invalidating" operation seems to have the ability to announce an object as "dead" while it is still reachable.  This is an interesting behaviour for GC.

But I still prefer the "marking" or "mark-and-move" function not having the object-graph-modifying behaviour like this, because a GC may scan an object many times for different purposes.  Marking, compacting, evacuating, doing reference-counting are just a few examples.  Anyway, it is an existing problem in Ruby, not a problem with this refactoring.

Reply:
I'm not sure what the best approach to solving this problem is, since there's a many-to-one relationship between the callcache (cc) and the callable_method_entry (cme). Because of this relationship, it's difficult to notify the cc that the cme has been invalidated when the cme becomes invalidated.

I made some changes in a37c47f971958f6b5fe521f4cf8c3ab6700b68a1 that combines the code paths for marking and reference updating. While this doesn't solve the problem of modifying the object graph during marking, it reduces the reliance of the code on the `reference_updating` argument.

Reply:
I have a thought that may be MMTk-specific, but may be applicable to Ruby's existing GC as well.  We can treating this field as a kind of "weak reference".

In MMTk, there is a phase for handling weak references after computing transitive closure (at which time the life or death of all objects are determined).  For Java, it will visit every `WeakReference` objects and, if the object it (weakly) points to is still alive, do nothing; otherwise, clear its reference field and "enqueue" that `WeakReference` (so that the application can process it further, such as clearing associated resources).  In my current mmtk-ruby binding, during that phase, I call `obj_free` for dead objects, and handle global weak tables such as `finalizer_table`, `obj_to_id_tbl`, `id_to_obj_tbl`, etc., clearing their invalid entries (keys or values being dead objects).

We can adopt this approach to `iseq`.  I think callcache objects are invalidated in the `obj_free` of some objects. After all dead objects have their `obj_free` called, we visit every `iseq` that are still alive, and see if its `cc` field still points to valid `callcache` objects.  If not, we clear it at this time (assign `rb_vm_empty_cc()` to it).

The down side is obviously a one-word-per-iseq overhead.  We need to register all `iseq` into an array (or any set data structure) so that they can be enumerated during the weak-referernce-processing phase, unless there is another way to enumerate all `iseq` instances in the Ruby VM.

For vanilla Ruby, the equivalent time to do this may be the end of `gc_update_references`, where those global weak tables are updated as well.  There should be alternative place to do it in the case of non-moving GC (mark-sweep).

Reply:
The problem with moving this to reference updating is that Ruby's compaction is optional and disabled by default, meaning that for most workloads, GC compaction will never run. Another issue is that GC compaction only runs on major GC, so it won't be run for minor GC.

When GC compaction is not enabled, the only time a live object will be inspected by the GC is during marking, so that's our only opportunity to modify that object. I'm not a huge fan of adding an array that stores references to all live iseqs, since it's not needed in the current GC implementation and will have overhead.